lib/runtime-core/src/trampoline_x64.rs

```diff
@@ -407,6 +418,43 @@ mod tests {
         assert_eq!(ret, 136);
     }
 
+    #[test]
+    fn test_trampolines_with_floating_point() {
+        unsafe extern "C" fn inner(n: *const CallContext, args: *const u64) -> u64 {
+            let n = n as usize;
+            let mut result: u64 = 0;
+            for i in 0..n {
```

Comment:
Sorry I'm confused – `n` was a pointer to a CallContext, so if it's at address 0x00005620d48999e8 then this loop will run 94699004729832 times?

Reply:
The `*const CallContext` is casted from the length of the parameter list as an integer in the following call to `add_callinfo_trampoline`, so it's not really a pointer.

Reply:
Will add a comment.

Reply:
OK, but you can't just go around passing integers as pointers. The compiler makes the assumption that the pointer is sufficiently aligned to point to a `CallContext`, which could lead it to conclude that the bottom bits are zero. All you do here is use it as a `usize` which will cast it, but if you were to pass, say, a bitfield through, the compiler would actually delete the tests of the bottom bits.

I don't know about Rust, but in C and C++ the rule is that all pointers must actually point to a valid object when created (whenever an expression of pointer type is evaluated). You're guaranteed that casting a pointer to an appropriately sized int and back to the same pointer type gets you the original pointer back, but that's a different property from being able to take an int, cast it to a pointer, and put it back to an int. For a crazy example that no compiler I know of does, if there's a value in the range [0, 4) the compiler could pack it in the same storage as the int.

If this is a temporary part of the workaround that will go away with the permanent fix, let me know. Casting this way doesn't seem to go wrong in practice, much.

Reply:
This is part of the test code and does not go into the compiled crate, so the worst case is one more failing test and we know something is wrong here.

Reply:
Oh no, that's never the worst case. :-) The worst case is something like "not only does this test pass when it should actually fail, it causes the other tests running in other threads to incorrectly pass when they should be failing". Anyway, I don't think we need to change this right now.